modelling/feat_engineer/feature_engineering.py

```diff
@@ -103,16 +160,36 @@ def rgb_to_ppg(df: pd.DataFrame, filter='band', qe_constants=(0.004989, 0.001193
                 b = firwin(order+1, cutoff, )
                 a = 1.0
 
-        # Apply the filter (and normalise)
-        for bid in _df[block_id].unique():
-            for field in tx_fields:
+    for bid in tqdm(_df[block_id].unique()):
+        for field in tx_fields:
+
+            # Apply the filter
+            if filter is not None:
                 _df.loc[_df[block_id] == bid, field] = filtfilt(
                     b, a, _df[_df[block_id] == bid][field])
 
-                # Scaling
-                if scale:
-                    _df.loc[_df[block_id] == bid, field] = MinMaxScaler(
-                    ).fit_transform(_df[_df[block_id] == bid][field].values.reshape(-1, 1))
+            # Clip outliers
+            if outlier_thresh is not None:
+                sig = _df[_df[block_id] == bid][field]
+                stdev = sig.values.std()
+                median = np.median(sig.values)
+                mean = sig.mean()
+                sig = np.where(sig < (outlier_thresh*stdev),
+                               sig, mean+outlier_thresh*stdev)
+                sig = np.where(sig < (-outlier_thresh*stdev),
+                               mean-outlier_thresh*stdev, sig)
+                _df.loc[_df[block_id] == bid, field] = sig
+
+            # Min-Max scaling
+            if scale:
+                _df.loc[_df[block_id] == bid, field] = MinMaxScaler(
+                ).fit_transform(_df[_df[block_id] == bid][field].values.reshape(-1, 1))
+
+            # Apply smoothing
+            if smoothing_window is not None:
+                smoothed = _df.loc[_df[block_id] ==
+                                   bid, field].rolling(2).mean()
+                _df.loc[_df[block_id] == bid, field] = smoothed
```

Comment:
Nice! Would be interesting to compare the effects of those on the model. I'll fix some parameters for later comparison.

Reply:
Yeah, we have a lot of things we could tweak and test its effects, but we probably need to constrain the scope as much as possible. Still, the clipping feels like it could be useful to eliminate weird artifacts in the data.